### PR TITLE
ENG-391 feat(canonical lesson): canonical lesson page

### PR DESCRIPTION
**Pull request by mantagen**
## Description

- adds new curriculum api queries and pages (lesson-overview-canonical, lesson-downloads-canonical)
  - they use the same MVs as lesson-overview and lesson-downloads
  - it groups instances of the same lesson together, adding a list of pathways in which the lesson is found
- Adds components (with storybook stories, [should match Figma](https://www.figma.com/file/3i0y2PHT6pECaGgvLhoN8z/Lesson-page-July-2023?node-id=2111%3A54147&mode=dev)) for:
  - `TagFunctional`
  - `LessonAppearsIn`
  - `LessonPathwayCard`
- organises the files a little into a `@/components/Lesson` directory
- amended some Avo events so that some of pathway fields are optional, and those changes are reflected in this PR 

## How to test
- go to https://deploy-preview-1891--oak-web-application.netlify.thenational.academy/teachers/lessons/masculinity-as-macbeth-s-hamartia
- you should see links to the lesson page within three units at the bottom
- got to https://deploy-preview-1891--oak-web-application.netlify.thenational.academy/teachers/lessons/transverse-waves
- you should see units separated by subject like:
image

### CODE REVIEW

**Review by benprotheroe — COMMENTED** [review]
havent we retired beta?
the link doesnt take you anywhere

**Comment by mantagen** [comment]
> havent we retired beta?
> the link doesnt take you anywhere

apologies @benprotheroe , updated the description

**Review by jack-skerrett-oak — CHANGES_REQUESTED** [review]
Morning :) 

This is my first time looking at this canonical URL stuff so please tell me where I am wrong in my assumptions!

- This PR has made the "Download all resources button go BIG". On the currently deployed OWA this is not the case
Screenshot 2023-09-25 at 09 00 16 (2)
- When tiers get involved, there are duplicated links
Screenshot 2023-09-25 at 09 00 46
- The unit in the image above (transverse waves) also appears in a similar set of physics units but here we only have links to the combined science units
- The transverse waves unit has a download all button when in the traditional pathway, but it doesn't here

File: src/common-lib/urls/urls.ts [hunk 1]
```diff
@@ -525,13 +542,25 @@ export const OAK_PAGES: {
...
   "lesson-downloads": createOakPageConfig({
     pathPattern:
       "/teachers/programmes/:programmeSlug/units/:unitSlug/lessons/:lessonSlug/downloads",
     analyticsPageName: "Lesson Download",
     configType: "internal",
     pageType: "lesson-downloads",
   }),
+  "lesson-downloads-canonical": createOakPageConfig({
+    pathPattern: "/teachers/lessons/:lessonSlug/downloads",
+    analyticsPageName: "Lesson Download", // @todo this should be "Lesson Download (Canonical)"?
```
Comment (mantagen) [inline src/common-lib/urls/urls.ts:560]: @verityroberts @ewan-wakeman do we want to distinguish the canonical pages using the "Page name" prop? Or are we happy for it to be called "Lesson Download" regardless of whether the user is in a specific pathway?

File: src/components/Lesson/LessonAppearsIn/LessonAppearsIn.stories.tsx [hunk 2]
```diff
@@ -0,0 +1,163 @@
+import React from "react";
+import { StoryFn, Meta } from "@storybook/react";
+
+import { LessonAppearsIn as Component } from "./LessonAppearsIn";
+
+import Flex from "@/components/Flex";
+import MaxWidth from "@/components/MaxWidth/MaxWidth";
+
+export default {
+  title: "Lesson/Lesson Appears In",
```
Comment (mantagen) [inline src/components/Lesson/LessonAppearsIn/LessonAppearsIn.stories.tsx:10]: Group with other components

File: src/common-lib/urls/urls.ts [hunk 3]
```diff
@@ -111,6 +115,15 @@ type LessonDownloadsLinkProps = {
     preselected: PreselectedDownloadType | null;
   };
 };
+export type LessonDownloadsCanonicalLinkProps = {
+  page: "lesson-downloads-canonical";
+  programmeSlug: null;
```
Comment (mantagen) [inline src/common-lib/urls/urls.ts:120]: remove these

**Review by jack-skerrett-oak — CHANGES_REQUESTED** [review]
Sorry, I've got more things I didn't notice :( 
- [x] Could "Lesson appears in" be changed to a H2 so that the page layout makes more sense (and then all the headings below that move up one)
- [x] The hover state is not what is shown in figma
 Screenshot 2023-09-26 at 10 32 58
- [x] I don't know the full context behind these canonical URLs, but it does seem a bit odd to have something like this https://deploy-preview-1891--oak-web-application.netlify.thenational.academy/teachers/lessons/formation-of-i-l-and-t. The breadcrumb is fully visible unlike the lessons that are across multiple programmes...

File: src/components/Lesson/LessonPathwayCard/LessonPathwayCard.stories.tsx [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+  ],
+};
+LessonPathwayCard.parameters = {
+  backgrounds: {
+    default: "dark",
+  },
+};
+
+// {
+//     unitSlug: string;
```
Comment (mantagen) [inline src/components/Lesson/LessonPathwayCard/LessonPathwayCard.stories.tsx:46]: Remove comment

File: src/components/Lesson/lesson.helpers.test.ts [hunk 5]
```diff
@@ -0,0 +1,320 @@
+import { groupLessonPathways } from "./lesson.helpers"; // Replace with the actual import path
```
Comment (mantagen) [inline src/components/Lesson/lesson.helpers.test.ts:1]: remove comment

File: src/components/Lesson/lesson.helpers.ts [hunk 6]
```diff
@@ -0,0 +1,270 @@
...
+          acc[key] = null;
+        }
+      });
+      return acc;
+    },
+    { ...nullPathway },
+  );
+};
+
+export const getLessonOverviewBreadCumb = ({
```
Comment (mantagen) [inline src/components/Lesson/lesson.helpers.ts:44]: typo

File: src/components/Lesson/lesson.helpers.ts [hunk 7]
```diff
@@ -0,0 +1,270 @@
...
+      oakLinkProps: {
+        page: "lesson-overview-canonical",
+        lessonSlug,
+      },
+      label: lessonTitle,
+      disabled,
+    };
+  }
+};
+export const getLessonDownloadsBreadCumb = ({
```
Comment (mantagen) [inline src/components/Lesson/lesson.helpers.ts:79]: typo

File: src/components/Lesson/lesson.types.ts [hunk 8]
```diff
@@ -0,0 +1,65 @@
...
+import { KeyLearningPoint } from "@/components/KeyLearningPoints/KeyLearningPoints";
+import { KeyWord } from "@/components/KeyWords/KeyWords";
+import {
+  ContentGuidance,
+  Equipment,
+} from "@/components/LessonRequirements/LessonRequirements";
+import { TeacherTip } from "@/components/TeacherTips/TeacherTips";
+import { LessonOverviewQuizData } from "@/node-lib/curriculum-api-2023/shared.schema";
+
+export type LessonBase = {
```
Comment (mantagen) [inline src/components/Lesson/lesson.types.ts:11]: import from schema

File: src/components/SubjectIconBrushBorders/SubjectIconBrushBorders.tsx [hunk 9]
```diff
@@ -10,7 +10,7 @@ import { OakColorName } from "../../styles/theme/types";
 import TagPromotional from "../TagPromotional";
 
 type SubjectIconBrushBoardersProps = Omit<CMSImageProps, "image"> & {
-  subjectSlug: string;
+  subjectSlug: string | null;
```
Comment (mantagen) [inline src/components/SubjectIconBrushBorders/SubjectIconBrushBorders.tsx:13]: either have this optional, or optionally render this component

File: src/node-lib/curriculum-api-2023/queries/lessonOverviewCanonical/lessonOverviewCanonical.gql [hunk 10]
```diff
@@ -0,0 +1,38 @@
...
+    lessonSlug
+    lessonTitle
+    programmeSlug
+    unitSlug
+    unitTitle
+    keyStageSlug
+    keyStageTitle
+    subjectSlug
+    subjectTitle
+    examboardSlug: examBoardSlug
```
Comment (mantagen) [inline src/node-lib/curriculum-api-2023/queries/lessonOverviewCanonical/lessonOverviewCanonical.gql:14]: `examBoard`

**Review by kimon-satan — APPROVED** [review]

File: src/common-lib/urls/urls.ts (hunk 1 above)
Comment (mantagen) [reply src/common-lib/urls/urls.ts:560]: For now I will leave them as is. Let me know if you want to differentiate in the tracking plan

**Review by Monika-Koziol — COMMENTED** [review]
There are a few differences between Figma and code. None of them to big to not go live with them

- [x] "Lesson appears on" should use heading 5 styling
- [ ] Top and bottom padding set to 100px
- [ ] Unit Title to be using style of Heading 6 Light
- [ ] It would be great if different exam boards tags could be different colour but if it add too much work then let's do it another time

**Review by k-huggs — APPROVED** [review]

**Review by jack-skerrett-oak — APPROVED** [review]
Lesson data is in an odd place so will make a note to test once we have a better grasp on published data

**Pull request merged**